### ISSUE TITLE
Can't install Erlang-24 due to openssl issues

### ISSUE THREAD

**Issue by donileo** [kerl/kerl#371]
Getting the following when installing erlang 24 via `asdf install erlang latest` on MacOS Big Sur 11.3.1:

```checking for OpenSSL header in /usr/local/opt/openssl@1.1... yes
checking for OpenSSL in /usr/local/opt/openssl@1.1... configure: error: static linking against crypto library disabled by user, but no dynamic library found in /usr/local/opt/openssl@1.1
ERROR: /Users/adonis/.asdf/plugins/erlang/kerl-home/builds/asdf_24.0/otp_src_24.0/lib/crypto/configure failed!
```

That’s with the --with-ssl option specified or not. I did run `asdf plugin update --all` prior to running the install erlang latest command above.

**Comment by 9point6** [comment]
On a slightly older version of Big Sur (11.2.3), but getting the same error. Tried reinstalling openssl via `brew reinstall openssl` and also manually specifying the `export KERL_CONFIGURE_OPTIONS="--with-ssl=$(brew --prefix openssl)"` flag

**Comment by 9point6** [comment]
On a slightly older version of Big Sur (11.2.3), but getting a similar error. 

```
checking for OpenSSL header in /usr/local/opt/openssl@1.1... yes
checking for OpenSSL in /usr/local/opt/openssl@1.1... configure: error: neither static nor dynamic crypto library found in /usr/local/opt/openssl@1.1
ERROR: /Users/USERNAME/.asdf/plugins/erlang/kerl-home/builds/asdf_24.0/otp_src_24.0/lib/crypto/configure failed!
```

Tried reinstalling openssl via `brew reinstall openssl` and also manually specifying the `export KERL_CONFIGURE_OPTIONS="--with-ssl=$(brew --prefix openssl)"` flag

**Comment by joseph-lozano** [comment]
Getting the same error. I tried reinstalling openssl and running with the KERL_CONFIGURE_OPTIONS flag and I got the same result.

**Comment by ronyclau** [comment]
_TL;DR: read the last part._

I encountered the very same problem. After inspecting both `config.log` and `otp_build_24.0.log`, I found there were some problems in deducing some environment variables in `lib/crypto/configure`, and ended up patching that file.

1. Environment variables `CPP` and `EGREP` couldn't be deduced correctly and ended up empty, which in turn broke the OpenSSL detection. I worked around that by exporting these two:
   ```bash
   export CPP=cpp EGREP=egrep
   ```
2. `DED_LD` was incorrectly passed in by the process that invoked `lib/crypto/configure`. 
   
   This can be verified by checking L7 of `config.log` under `lib/crypto/configure`. Mine reads:
   ```
   $ ./configure CC=clang CPP=cpp DED_LD=clang 'DED_LDFLAGS=-m64 -bundle -bundle_loader /Users/USERNAME/.asdf/plugins/erlang/kerl-home/builds/asdf_24.0/otp_src_24.0/bin/x86_64-apple-darwin19.6.0/beam.smp' --enable-darwin-64bit --with-ssl=/usr/local/opt/openssl@1.1 --disable-option-checking --cache-file=/dev/null --srcdir=/Users/USERNAME/.asdf/plugins/erlang/kerl-home/builds/asdf_24.0/otp_src_24.0/lib/crypto
   ```

   This disabled a special handling of `DED_*` flags for OSX, as the logic only executes if `DED_LD` was empty. 

   `lib/crypto/configure` L4883-L4885:
   ```bash
   # If DED_LD is set in environment, we expect all DED_LD* variables
   # to be specified (cross compiling)
   if test "x$DED_LD" = "x"; then
   ```
   `lib/crypto/configure` L4907-L4914:
   ```bash
	darwin*)
		# Mach-O linker: a shared lib and a loadable
		# object file is not the same thing.
		DED_LDFLAGS="-bundle -bundle_loader ${ERL_TOP}/bin/$host/beam.smp"
		# DED_LDFLAGS_CONFTEST is for use in configure tests only. We
		# cannot use DED_LDFLAGS in configure tests since beam.smp has not
		# been built yet...
		DED_LDFLAGS_CONFTEST="-bundle"
   ```

   I couldn't override that by `export`-ing an empty `DED_LD`, and ended up patching `lib/crypto/configure`. I was using `asdf`, so you may have to adjust the path:
   ```bash
   printf "1a\nunset DED_LD\n.\n\nw\nq\n" | ed ~/.asdf/plugins/erlang/kerl-home/builds/asdf_24.0/otp_src_24.0/lib/crypto/configure
   ```

   Since kerl` always re-extract the source package on compilation, you have to wait until the extraction process is finished before invoking the command.

---
My workaround assumes that you are using `asdf` (which in turn invokes `kerl`) to install `erlang`. You may need to adjust the command/path if you are directly using `kerl`.

First, open two terminals, or pre-create a split in your favourite terminal multiplexer.

Run this in one terminal:
```bash
$ CPP=cpp EGREP=egrep asdf install erlang latest:24
```

Wait until this line appears:
```
Building Erlang/OTP 24.0 (asdf_24.0), please wait...
```

Then, quickly in another terminal:
```bash
# Make sure the file is created, if not, try again a couple seconds later.
$ ls ~/.asdf/plugins/erlang/kerl-home/builds/asdf_24.0/otp_src_24.0/lib/crypto/configure

# Patch `configure`
$ printf "1a\nunset DED_LD\n.\n\nw\nq\n" | ed ~/.asdf/plugins/erlang/kerl-home/builds/asdf_24.0/otp_src_24.0/lib/crypto/configure
```

If the patch is applied successfully, these few lines should appear (the numbers may be different on your side):
```
224018
# Guess values for system-dependent variables and create Makefiles.
224031
```

The build should continue as expected.

**Comment by t-rutten** [comment]
Thanks for workaround @ronyclau! Seems to work for `asdf`, but has anyone has success with it applied to `kerl` directly? I'm trying

```
kerl build 24.0 24.0
```
then in the other terminal
```
printf '/if test "x$DED_LD" = "x"; then\ni\nunset DED_LD\n.\n\nw\nq\n' | ed .kerl/builds/24.0/otp_src_24.0/lib/crypto/configure
```
I see the "patch successful" lines, but the build fails with the same `configure: error: neither static nor dynamic crypto library` as before.

**Comment by ronyclau** [comment]
@t-rutten Did you export or prefix the command with `CPP=cpp EGREP=egrep`?

That is:
```bash
export CPP=cpp EGREP=egrep
kerl build 24.0 24.0
```

Or
```bash
CPP=cpp EGREP=egrep kerl build 24.0 24.0
```

**Comment by kamidev** [comment]
Thanks @ronyclau! The patch works for me on an Intel Mac running 10.15.7. I used asdf, haven't tried applying kerl directly.

**Comment by ronyclau** [comment]
This seems to be related to https://github.com/erlang/otp/issues/4821

**Comment by kamidev** [comment]
> This seems to be related to [erlang/otp#4821](https://github.com/erlang/otp/issues/4821)

Yes. And someone there suggests there may be [multiple problems.](https://github.com/erlang/otp/issues/4821#issuecomment-840311958) causing the issue.

When I successfully applied your patch, I installed like this: EGREP=egrep CC=clang CPP="clang -E" KERL_USE_AUTOCONF=0 KERL_CONFIGURE_OPTIONS="--with-ssl=$(brew --prefix openssl) --enable-wx" asdf install erlang 24.0

**Comment by jj1bdx** [comment]
FYI: you can suspend a process with CTRL/Z (^z) before patching `configure`. A great hack anyway! :)

**Comment by paulo-ferraz-oliveira** [comment]
Downgrading `autoconf` to 2.69 did it for me. Still trying to understand if this is an OTP specific issue: https://github.com/erlang/otp/issues/4821#issuecomment-840154032.

**Comment by paulo-ferraz-oliveira** [comment]
https://github.com/erlang/otp/issues/4821#issuecomment-840342780 worked for me, but I'm unsure of the implications of bypassing the flags. Downgrading `autoconf` also made my compilation flow work (again!) for 21.3.8.23 and 22.3.4.18.

**Comment by cjntaylor** [comment]
> Wait until this line appears:
> 
> ```
> Building Erlang/OTP 24.0 (asdf_24.0), please wait...
> ```
> 
> Then, quickly in another terminal:
> 
> ```shell
> # Patch `configure`
> # For asdf
> $ printf '/if test "x$DED_LD" = "x"; then\ni\nunset DED_LD\n.\n\nw\nq\n' | ed ~/.asdf/plugins/erlang/kerl-home/builds/asdf_24.0/otp_src_24.0/lib/crypto/configure
> 
> # For kerl
> $ printf '/if test "x$DED_LD" = "x"; then\ni\nunset DED_LD\n.\n\nw\nq\n' | ed ~/.kerl/builds/24.0/otp_src_24.0/lib/crypto/configure
> ```
> 
> If you see `No such file or directory`, the file may yet to have generated by the build process. Try again a couple of seconds later.
> 
> If the patch is applied successfully, these few lines should appear (the numbers may be different on your side):
> 
> ```
> 224018
> if test "x$DED_LD" = "x"; then
> if test "x$DED_LD" = "x"; then
> 224031
> ```
> 
> The build should continue as expected.

Oh no! Please, please don't suggest something like this. You're relying on a race condition to work properly - that `kerl` has finished wiping away the build directory, unpacking the sources and running `autoconf` before you make the change. This isn't repeatable at all, or reliable at all. If you make the change at the wrong time, the build will still fail.

Furthermore, and this is the important bit ***this only fixes crypto***. The OTP developers have pointed out that this is a more generalized problem with Autoconf > 2.69: https://github.com/erlang/otp/issues/4821#issuecomment-840196135

This patch note in Autoconf 2.70 indicates that they changed ***in general*** how macro dependencies work, and how certain macros parse. The impact of this is unknown right now. Yes, this "fix" may allow `24.0` to build. However, there are likely other parts of the build that are getting misconfigured or built incorrectly due to the Autoconf macros not parsing the same way. The only correct, reliable way to fix this is to build with Autoconf 2.69 (https://github.com/erlang/otp/issues/4821#issuecomment-840154032)

**Comment by paulo-ferraz-oliveira** [comment]
[As per @rickard-green](https://github.com/erlang/otp/issues/4821#issuecomment-845444243), the "solution should be to not set any DED_* variables on MacOS. It should work out of the box. It is possible to set `DED_*` variables also on MacOS, but then you also need to set the `DED_LDFLAGS_CONFTEST` variable which will be used in the link configure tests. See the [important variables inspected by configure](https://github.com/erlang/otp/blob/master/HOWTO/INSTALL.md#important-variables-inspected-by-configure) section of `HOWTO/INSTALL.md`."

Then [there's also this comment](https://github.com/erlang/otp/issues/4821#issuecomment-845558118), by the same author, which seems relevant.

To make this work with OTP 24.0 I didn't have to downgrade to `autoconf` 2.69, anymore, but did have to apply [@jj1bdx's patch](https://github.com/erlang/otp/issues/4821#issuecomment-840342780) to `kerl` so something like it might be in order.

**Comment by optikfluffel** [comment]
@paulo-ferraz-oliveira have you tried the new `24.0.1` release? This one compiled fine on my Mac without any special hackery, patches or anything.

**Comment by paulo-ferraz-oliveira** [comment]
I did. I was so hopeful I even tweeted about it, to then test without the `kerl` patch and see it fail. There's another variable here, of course, which is MacOS version :-); yours might not be affected.

**Comment by optikfluffel** [comment]
@paulo-ferraz-oliveira yes maybe, I'm still running Catalina (`10.15.7`) over here ^^

**Comment by paulo-ferraz-oliveira** [comment]
Here's some more info. so we can get a patch out soon: https://github.com/erlang/otp/issues/4821#issuecomment-845914942.

**Comment by michaelklishin** [comment]
Erlang `24.0.1` is out. I've successfully built it on both GA and beta versions of Big Sur.

**Comment by yukster** [comment]
24.0.1 still fails with the OpenSSL issue for me on Catalina 10.15.7.

**Comment by bencoppock** [comment]
@yukster it sounds like 24.0.1 fixed one of two issues, so you no longer need to downgrade autoconf to compile the latest Erlang. However, you still need to follow [step 2 here](https://github.com/erlang/otp/issues/4821#issuecomment-840342780).

It took me a minute to find the file that comment wants you to patch. In my case, since I'm using asdf, it ended up being in ~/.asdf/plugins/erlang/kerl. Then I searched for the lines mentioned in that comment and applied the changes. After that compilation on Catalina worked for me.

**Comment by yukster** [comment]
Thanks @bencoppock! I'm not in too big a rush to get on 24 (tho I am super excited for it) so I'll just wait for a revision that works. :-)

**Comment by eduardodeoh** [comment]
Quick fix: https://www.youtube.com/watch?v=w7JkhGrjnMY

**Comment by akash-akya** [comment]
Just setting `DED_LD` to empty string worked for me for erlang `v24.0.1`. Didn't do any other changes.

```bash
DED_LD=" " asdf install erlang 24.0.1
```

Setting it to empty string makes sure that kerl does not set this variable and then it appears`otp_build` takes care of removing empty env variables

**Comment by mrallen1** [comment]
@akash-akya sounds like the rule would be something like

* if os is Big Sur and the Erlang build is 24, then unset `DED_LD`

Is that correct?

**Comment by akash-akya** [comment]
Its not specific to Big Sur, since mine is Catalina. I think it might have to do with the clang version (mine is 11) but don't know for sure. 
```
λ clang --version
Apple clang version 11.0.3 (clang-1103.0.32.62)
Target: x86_64-apple-darwin19.5.0
Thread model: posix
InstalledDir: /Library/Developer/CommandLineTools/usr/bin
```

Also ideally we should [unset all](https://erlang.org/doc/installation_guide/INSTALL-CROSS.html#cross-compiler-and-other-tools) `DED_LD_*` variables.

**Comment by mrallen1** [comment]
Good to know. I switched to Ubuntu when Apple took away my Esc key (thinking about moving back to Macs now that Esc is back) - anyway this is difficult for me to diagnose at a distance - so guidance from people who _are_ using Macs as their primary platforms is so useful.

Thanks.

**Comment by akash-akya** [comment]
It seems issue also exists in [mojave](https://github.com/erlang/otp/issues/4821#issuecomment-845368632).

I think we can just skip setting these variables for recent OTP releases? OTP 20.1+ based on [this comment](https://github.com/erlang/otp/issues/4821#issuecomment-845914942)

**Comment by yukster** [comment]
> Just setting `DED_LD` to empty string worked for me for erlang `v24.0.1`. Didn't do any other changes (as previously mentioned 24.0.1 fixed one out of two issues).
> 
> ```shell
> DED_LD=" " asdf install erlang 24.0.1
> ```
> 
> Setting it to empty string makes sure that kerl does not set this variable and then it appears`otp_build` takes care of removing empty env variables

This worked for me on both Catalina and Mojave! Thanks @akash-akya!

**Comment by mrallen1** [comment]
@akash-akya I'm going to try and get this fixed up and push out a new release in a day or two here. Thanks for the link to that OTP comment.

**Comment by mrallen1** [comment]
Would love if a macOS user runs my branch and does some test builds to see if this fixes things

**Comment by jj1bdx** [comment]
On macOS 10.15.7 + Xcode 12.4, the following test build went fine. (Note: jj1bdx/otp is equivalent to erlang/otp)

```
% env \
    CC=/usr/bin/clang \
    CXX=/usr/bin/clang++ \
    CFLAGS="-O3 -fstack-protector-strong" LDFLAGS="-fstack-protector-strong" \
    MAKEFLAGS="-j12" \
    KERL_BUILD_PLT="yes" \
    KERL_CONFIGURE_OPTIONS="--enable-m64-build --enable-darwin-64bit --disable-hipe --enable-vm-probes --with-dynamic-trace=dtrace --disable-native-libs --enable-kernel-poll --enable-dirty-schedulers --enable-sharing-preserving --enable-lock-counter --enable-esock --disable-sctp --without-javac --without-odbc --enable-threads --enable-smp-support --with-wx --with-ssl=/usr/local/opt/openssl --disable-silent-rules --disable-dynamic-ssl-lib" \
    KERL_BUILD_DOCS="yes" \
sh -c "(kerl build git https://github.com/jj1bdx/otp/ OTP-24.0.2 test-24.0.2)"
Checking out Erlang/OTP git repository from https://github.com/jj1bdx/otp/...
Building Erlang/OTP test-24.0.2 from git, please wait...
APPLICATIONS DISABLED (See: /Users/kenji/.kerl/builds/test-24.0.2/otp_build_git.log)
 * jinterface     : Java compiler disabled by user
 * odbc           : User gave --without-odbc option

DOCUMENTATION INFORMATION (See: /Users/kenji/.kerl/builds/test-24.0.2/otp_build_git.log)
 * documentation  :
 *                  fop is missing.
 *                  Using fakefop to generate placeholder PDF files.

Building docs...
Erlang/OTP test-24.0.2 from git has been successfully built
```

**Comment by mrallen1** [comment]
Cool thanks @jj1bdx

**Comment by jeregrine** [comment]
@mrallen1 this worked for me as well.

**Comment by mrallen1** [comment]
OK, I will merge and push a release. Thanks for your help

**Comment by adamu** [comment]
For any asdf users reading this far.

```
asdf plugin-update erlang
asdf install erlang 24.0.2
```

Thanks to everybody for working on solving the problems on OSX.

**Comment by zachdaniel** [comment]
updating the erlang plugin doesn't resolve the issue for me :(

**Comment by toranb** [comment]
@zachdaniel I fought w/ this for a good 2 days and this .zshrc setup worked for me

```export KERL_CONFIGURE_OPTIONS="--without-javac --with-ssl=/opt/homebrew/Cellar/openssl@1.1/1.1.1l"

source /opt/homebrew/opt/asdf/libexec/asdf.sh```

If that doesn't work, try this (another solid option worth looking into)

`asdf install erlang 24.0.6 --with-ssl=$(brew --prefix openssl)`

**Comment by HangingClowns** [comment]
on M1 macs i keep getting:

```
Build failed.
                        ^
/opt/homebrew/opt/openssl@3/include/openssl/macros.h:62:52: note: expanded from macro 'OSSL_DEPRECATED'
#   define OSSL_DEPRECATED(since) __attribute__((deprecated))
                          ^
10 warnings generated.
25 warnings generated.
make[3]: *** [opt] Error 2
make[2]: *** [opt] Error 2
make[1]: *** [opt] Error 2
make: *** [libs] Error 2
```

**Comment by toranb** [comment]
@HangingClowns I had to `rm -rf openssl@3` and `brew docker` (running any prune /symlink necessary) because having this newer version didn't work w/ my M1

**Comment by HangingClowns** [comment]
so then we need to use `openssl@1.1`?

**Comment by toranb** [comment]
@HangingClowns yup - this was exactly what I used on my M1 tonight (in the home dir .zshrc)

`export KERL_CONFIGURE_OPTIONS="--without-javac --with-ssl=/opt/homebrew/Cellar/openssl@1.1/1.1.1l"`

**Comment by HangingClowns** [comment]
this works on Mac with Intel: `export KERL_CONFIGURE_OPTIONS="--without-javac --with-ssl=/usr/local/opt/openssl@1.1" `

**Comment by kaka-ruto** [comment]
None of the above worked for me on an M1 Mac. Installing `Erlang 24.2` worked so I'll upgrade to that instead

**Comment by guisfits** [comment]
I had the same problem, I solved uninstalling `wxwidgets`

**Comment by jdmarshall** [comment]
Still not working with libressl on OS X 

configure: error: Invalid path to option --with-ssl=PATH

using --with-ssl=/usr/bin/openssl

**Comment by jj1bdx** [comment]
I understand that macOS /usr/bin/openssl has no associated include files. I guess you need to install an OpenSSL separately.

**Comment by jadeallenx** [comment]
> Still not working with libressl on OS X
> 
> configure: error: Invalid path to option --with-ssl=PATH
> 
> using --with-ssl=/usr/bin/openssl

You need to point to the LibreSSL development headers/library stuff, not the commandline tool(s). I am not sure where those might be though.

**Comment by GPrimola** [comment]
Just for the record, I was having a bad time with this issue in the past 3 days.
Unfortunately none of the suggestions here helped me, but when I was about to re my whole computer to start it from scratch, I figured my Terminal.app was `open using Rosetta` and after I started it in normal mode (not using Rosetta), I could finally have Erlang installed (with the --with-ssl=$(brew --prefix openssl@1.1) flag)!

My guess is the openssl lib installed in my computer is for `arm64`, but while I was using Terminal.app on Rosetta, Erlang was being compiled to `x86_x64` arch.

I hope this can also be a clue for other people (or even myself again) in the future.

**Comment by jakecurreri** [comment]
Another workaround is to use Rosetta to emulate an `x86_64` environment for the Erlang installation:

```shell
arch -x86_64 zsh
asdf install erlang latest
```

When you're done, exit terminal and open a new window to return `arm64`.

**Comment by wxygeek** [comment]
For elang with version under 24, the following steps work for me:

```
brew install openssl@1.1
brew link openssl@1.1
asdf install erlang 23.3.4.5
```

**Comment by Napolskih** [comment]
% brew install openssl@1.1
Error: openssl@1.1 has been disabled because it is not supported upstream! It was disabled on 2024-10-24.

What can be done?

**Comment by jj1bdx** [comment]
I guess you're still able to obtain a copy of OpenSSL 1.1.x, for example at: https://github.com/openssl/openssl/releases/tag/OpenSSL_1_1_1w

**Comment by mirrec** [comment]
None of the above solutions worked for me. But I was able to fix the issue with [solution mentioned elsewhere](https://github.com/asdf-vm/asdf-erlang/issues/319#issuecomment-2459115601).
 
```
$ KERL_CONFIGURE_OPTIONS="--disable-jit" asdf install erlang 27.2
```

I did update erlang plugins before, so the following command might be needed as well.

```
asdf plugin-update erlang
```
 
 Versions:
- erlang 27.2
- asdf 0.15.0
- OpenSSL 3.4.0
- macOS 15.2 (Sequoia)
